rand_distr/src/weighted_tree.rs

```diff
@@ -0,0 +1,381 @@
+        for weight in subtotals.iter() {
+            if *weight < W::ZERO {
+                return Err(WeightedError::InvalidWeight);
+            }
+        }
+        let n = subtotals.len();
+        for i in (1..n).rev() {
+            let w = subtotals[i].clone();
+            let parent = (i - 1) / 2;
+            subtotals[parent]
+                .checked_add_assign(&w)
+                .map_err(|()| WeightedError::Overflow)?;
+        }
+        Ok(Self { subtotals })
+    }
+
+    /// Returns `true` if the tree contains no weights.
+    pub fn is_empty(&self) -> bool {
+        self.subtotals.is_empty()
+    }
+
+    /// Returns the number of weights.
+    pub fn len(&self) -> usize {
+        self.subtotals.len()
+    }
+
+    /// Returns `true` if we can sample.
+    ///
+    /// This is the case if the total weight of the tree is greater than zero.
+    pub fn is_valid(&self) -> bool {
+        if let Some(weight) = self.subtotals.first() {
+            *weight > W::ZERO
+        } else {
+            false
+        }
+    }
+
+    /// Gets the weight at an index.
+    pub fn get(&self, index: usize) -> W
+    where
+        W: for<'a> SubAssign<&'a W>,
+    {
+        let left_index = 2 * index + 1;
+        let right_index = 2 * index + 2;
+        let mut w = self.subtotals[index].clone();
+        w -= &self.subtotal(left_index);
+        w -= &self.subtotal(right_index);
+        w
+    }
+
+    /// Removes the last weight and returns it, or [`None`] if it is empty.
+    pub fn pop(&mut self) -> Option<W>
+    where
+        W: for<'a> SubAssign<&'a W>,
+    {
+        self.subtotals.pop().map(|weight| {
+            let mut index = self.len();
+            while index != 0 {
+                index = (index - 1) / 2;
+                self.subtotals[index] -= &weight;
+            }
+            weight
+        })
+    }
+
+    /// Appends a new weight at the end.
+    pub fn push(&mut self, weight: W) -> Result<(), WeightedError> {
+        if weight < W::ZERO {
+            return Err(WeightedError::InvalidWeight);
+        }
+        if let Some(total) = self.subtotals.first() {
+            let mut total = total.clone();
+            if total.checked_add_assign(&weight).is_err() {
+                return Err(WeightedError::Overflow);
+            }
+        }
+        let mut index = self.len();
+        self.subtotals.push(weight.clone());
+        while index != 0 {
+            index = (index - 1) / 2;
+            self.subtotals[index].checked_add_assign(&weight).unwrap();
+        }
+        Ok(())
+    }
+
+    /// Updates the weight at an index.
+    pub fn update(&mut self, mut index: usize, weight: W) -> Result<(), WeightedError>
+    where
+        W: for<'a> SubAssign<&'a W>,
+    {
+        if weight < W::ZERO {
+            return Err(WeightedError::InvalidWeight);
+        }
+        let mut difference = weight;
+        difference -= &self.get(index);
```

Comment:
This will panic (in debug builds) with unsigned types when the new weight is less than the old. I'm not sure what the best solution is.
1. Just add an advisory notice?
2. Use wrapping arithmetic? That's probably not viable due to our generics & traits.
3. Use two paths, where the difference is less and where it's greater? (We already have a third, where there is no difference. Also, it *might* optimise out to use wrapping arithmetic on unsigned ints.)
4. Subtract then add? More expensive.

It's ugly, but (3) looks like the best option?